File: ROS/src/control/MPC/pkgs/environments/bicycle_model.py
import numpy as np
from environments.Env import Env
from environments.env_utils import Space


class BicycleModel(Env):
    def __init__(self, dt=0.01, R=0.2, L=1.6, Lr=0.7, mu=0.1, DC=0.025):
        self.dt = dt
        self.R = R
        self.L = L
        self.Lr = Lr
        self.mu = mu
        self.DC = DC

        self.action_space = Space(low=[-np.inf, -np.inf], high=[-np.inf, np.inf])

        high_state_space = np.array([np.inf, np.inf, np.inf, np.inf, np.inf])
        self.state_space = Space(low=-high_state_space, high=high_state_space)

        high_obs_space = np.array([np.inf, np.inf, np.inf, np.inf, np.inf])
        self.observation_space = Space(low=-high_obs_space, high=high_obs_space)

        self.F = self.discretize(self.dynamics, self.dt, M=1)

    def dynamics(self, s, u):
        # States
        # x = s[0]
        # y = s[1]
        theta = s[2]
        zeta = s[3]
        v = s[4]

        # Inputs
        alpha = u[0]
        phi = u[1]

        # Outputs
        a = alpha * self.R

        omega = v * np.tan(zeta) / self.L

        dx = v * np.cos(theta)
        dy = v * np.sin(theta)

        dv = a

        return np.array([dx, dy, omega, phi, dv])

    def reset(self):
        self.a = 0
        self.v = 0

        self.x = 0
        self.y = 0
        self.theta = 0
        self.zeta = 0
        self.ang_vel = 0
        self.phi = 0
        self.omega = 0

    def obs(self, x):
        return np.sqrt((x[0] - x[1]) ** 2 + (x[2] - x[3]) ** 2)

File: ROS/src/common/ugr_ros_control/sim_control/src/bicycle_model.cpp

#include <sim_control/bicycle_model.hpp>
#include <rosparam_shortcuts/rosparam_shortcuts.h>
#include <math.h>
#include <tuple>
#include <ros/ros.h>

BicycleModel::BicycleModel(double R, double L, double Lr, double mu, double DC) : R(R), L(L), Lr(Lr), mu(mu), DC(DC)
{
  this->reset();
};

void BicycleModel::reset()
{
  this->stop();

  this->x = 0;
  this->y = 0;
  this->theta = 0;
  this->zeta = 0;
  this->alpha = 0;
  this->ang_vel = 0;
  this->phi = 0;
  this->omega = 0;
};

void BicycleModel::stop()
{
  this->a = 0;
  this->phi = 0;
  this->v = 0;
};

void BicycleModel::update(double dt, double in_alpha, double in_phi)
{
  if (fabs(dt) < 0.0001)
  {
    return;
  }
  ROS_INFO_STREAM_THROTTLE(0.05, "BicycleModel::update: dt: " << dt << " in_alpha: " << in_alpha << " in_phi: " << in_phi);

  // Inputs
  alpha = in_alpha;
  phi = in_phi;

  // Angular velocity (of wheel) and steering angle

  double drag_acc = v > 0.1 ? DC * pow(v, 2) : 0.0;
  double friction_acc = v > 0.1 ? mu : 0.0;

  // drag_acc = 0.0;
  // friction_acc = 0.0;

  a = alpha * R; //- drag_acc - friction_acc;

  

  // Outputs (and intermediates)
  omega = v * tan(zeta) / L;

  double x_vel = v * cos(theta);
  double y_vel = v * sin(theta);

  x += x_vel * dt;
  y += y_vel * dt;
  v += a * dt;          // v of CoG (+ drag + friction)
  ang_vel = v / R;
  theta += omega * dt;  // Heading
  zeta += phi * dt;     // Of STEERING JOINT
  // ROS_INFO_STREAM("BicycleModel::update: delta_x: " << x_vel * dt << " delta_y: " << y_vel * dt << " delta_v: " << a * dt << " delta_theta: " << omega * dt << " delta_zeta: " << phi * dt);
  // ROS_INFO_STREAM("zeta: " << zeta);
  // zeta = std::min(std::max(zeta, -M_PI / 4), M_PI / 4);
};